crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -177,39 +173,45 @@ impl SettlementEncoder {
             .executed_amounts(*sell_price, *buy_price)
             .context("impossible trade execution")?;
 
-        self.pre_interactions
-            .append(&mut order.interactions.pre.clone());
         self.order_trades.push(order_trade);
         Ok(execution)
     }
 
-    // Fails if any used sell-token doesn't have a price. or if executed amount is impossible.
-    pub fn add_liquidity_order_trade(
+    /// Adds the passed trade to the execution plan. Handles specifics of ordinary, limit and
+    /// liquidity orders internally.
+    pub fn add_trade(
         &mut self,
         order: Order,
         executed_amount: U256,
         scaled_unsubsidized_fee: U256,
     ) -> Result<TradeExecution> {
         verify_executed_amount(&order, executed_amount)?;
-        // For the encoding strategy of liquidity orders, the sell prices are taken from
-        // the uniform clearing price vector. Therefore, either there needs to be an existing price
-        // for the sell token in the uniform clearing prices or we have to create a new price entry beforehand,
-        // if the sell token price is not yet available
-        if self.token_index(order.data.sell_token) == None {
-            let sell_token = order.data.sell_token;
-            let sell_price = order.data.buy_amount;
-            self.tokens.push(sell_token);
-            self.clearing_prices.insert(sell_token, sell_price);
-            self.sort_tokens_and_update_indices();
+        let interactions = order.interactions.clone();
+        let execution = match order.metadata.class {
+            OrderClass::Ordinary => {
+                self.add_ucp_trade(order, executed_amount, scaled_unsubsidized_fee)?
+            }
+            OrderClass::Liquidity => {
+                let buy_price = self.compute_limit_buy_price(&order)?;
+                self.add_custom_price_trade(
+                    order,
+                    executed_amount,
+                    scaled_unsubsidized_fee,
+                    buy_price,
+                )?
+            }
+            OrderClass::Limit => {
+                todo!()
```

Comment:
Should this be `bail!` just because it is technically possible to add limit orders in staging?

Reply:
Yeah, makes sense. 👍

Reply:
Just came back to say that this goes away in #713, so `todo!()` was fine as well.

Reply:
But it's nicer anyway to have PRs that are correct on their own.